homeassistant/auth/__init__.py

```diff
@@ -100,20 +102,20 @@ def __init__(self, hass: HomeAssistant, auth_manager: AuthManager) -> None:
 
     async def async_create_flow(
         self,
-        handler_key: str,
+        handler_key: tuple[str, str],
         *,
         context: dict[str, Any] | None = None,
         data: dict[str, Any] | None = None,
-    ) -> data_entry_flow.FlowHandler:
+    ) -> data_entry_flow.BaseFlowHandler[AuthFlowResult, tuple[str, str]]:
         """Create a login flow."""
         auth_provider = self.auth_manager.get_auth_provider(*handler_key)
         if not auth_provider:
             raise KeyError(f"Unknown auth provider {handler_key}")
         return await auth_provider.async_login_flow(context)
 
     async def async_finish_flow(
-        self, flow: data_entry_flow.BaseFlowHandler, result: FlowResult
-    ) -> FlowResult:
+        self, flow: data_entry_flow.BaseFlowHandler, result: AuthFlowResult
```

Comment:
```suggestion
        self, flow: LoginFlow, result: AuthFlowResult
```
Does this work? It would make the `cast` unnecessary.

Reply:
It doesn't work because:
```
homeassistant/auth/__init__.py:118: error: Argument 1 of "async_finish_flow" is incompatible with supertype "FlowManager"; supertype defines the argument type as "FlowHandler[AuthFlowResult, tuple[str, str]]"  [override]
homeassistant/auth/__init__.py:118: note: This violates the Liskov substitution principle
homeassistant/auth/__init__.py:118: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
```

I think mypy is being a bit unreasonable here though.
One option could be to provide the type of `FlowHandler` as a generic parameter to `FlowManager`.

Reply:
> One option could be to provide the type of `FlowHandler` as a generic parameter to `FlowManager`.

Wouldn't that also require higher-kind TypeVars as `FlowHandler` itself is generic?
I'd just leave it as is. Although nice, the current typing doesn't hurt and it's just one `cast`.